kernel/src/transaction.rs

```diff
@@ -0,0 +1,127 @@
+use crate::EngineInterface;
+use crate::{actions::Transaction, DeltaResult};
+
+pub use crate::actions::visitors::TransactionMap;
+pub struct TransactionScanner {
+    snapshot: Arc<Snapshot>,
+}
+
+impl TransactionScanner {
+    pub fn new(snapshot: Arc<Snapshot>) -> Self {
+        TransactionScanner { snapshot }
+    }
+
+    /// Scan the entire log for all application ids but terminate early if a specific application id is provided
+    fn scan_application_transactions(
+        &self,
+        engine: &dyn EngineInterface,
+        application_id: Option<&str>,
+    ) -> DeltaResult<TransactionMap> {
+        let schema = get_log_schema().project(&[TRANSACTION_NAME])?;
+
+        let mut visitor = TransactionVisitor::new(application_id.map(|s| s.to_owned()));
+
+        // when all ids are requested then a full scan of the log to the latest checkpoint is required
+        let iter =
+            self.snapshot
+                .log_segment
+                .replay(engine, schema.clone(), schema.clone(), None)?;
+
+        for maybe_data in iter {
+            let (txns, _) = maybe_data?;
+            txns.extract(schema.clone(), &mut visitor)?;
+            // if a specific id is requested and a transaction was found, then return
+            if application_id.is_some() && !visitor.transactions.is_empty() {
+                break;
+            }
+        }
+
+        Ok(visitor.transactions)
+    }
+
+    /// Scan the Delta Log for the latest transaction entry of an application
+    pub fn application_transaction(
+        &self,
+        engine: &dyn EngineInterface,
+        application_id: &str,
+    ) -> DeltaResult<Option<Transaction>> {
+        let mut transactions = self.scan_application_transactions(engine, Some(application_id))?;
+        Ok(transactions.remove(application_id))
+    }
+
+    /// Scan the Delta Log to obtain the latest transaction for all applications
+    pub fn application_transactions(
+        &self,
+        engine: &dyn EngineInterface,
+    ) -> DeltaResult<TransactionMap> {
+        self.scan_application_transactions(engine, None)
+    }
+}
+
+#[cfg(all(test, feature = "default-client"))]
+mod tests {
+    use std::path::PathBuf;
+
+    use super::*;
+    use crate::client::sync::SyncEngineInterface;
+    use crate::Table;
+
+    fn get_latest_transactions(path: &str, app_id: &str) -> (TransactionMap, Option<Transaction>) {
+        let path = std::fs::canonicalize(PathBuf::from(path)).unwrap();
+        let url = url::Url::from_directory_path(path).unwrap();
+        let engine_interface = SyncEngineInterface::new();
+
+        let table = Table::new(url);
+        let snapshot = table.snapshot(&engine_interface, None).unwrap();
+        let txn_scan = TransactionScanner::new(snapshot.clone());
+
+        (
+            txn_scan
+                .application_transactions(&engine_interface)
+                .unwrap(),
+            txn_scan
+                .application_transaction(&engine_interface, &app_id)
```

Comment:
another [clippy failure](https://github.com/delta-incubator/delta-kernel-rs/actions/runs/8862391787/job/24335971523?pr=170):
```
error: this expression creates a reference which is immediately dereferenced by the compiler
  --> kernel/src/transaction.rs:88:61
   |
88 |                 .application_transaction(&engine_interface, &app_id)
   |                                                             ^^^^^^^ help: change this to: `app_id`
```
(probably worth running `cargo clippy` locally to catch any other issues, since the build seems to only report one at a time)

Reply:
yeah my bad. I forgot to add the `--tests` flag when running. I'll use the just file going forwards